src/ocean/util/container/map/TreeMap.d

```diff
@@ -0,0 +1,352 @@
+
+        Returns:
+            true if the map is empty or false if it contains elements.
+
+    ***************************************************************************/
+
+    public bool is_empty ( ) // const
+    {
+        return this.root.is_empty();
+    }
+
+    /***************************************************************************
+
+        Reinitialises the internal tree struct.
+
+        This method should be called only if it is known that the tree is empty,
+        otherwise pointers to malloc-allocated nodes are lost so they cannot be
+        deallocated.
+
+    ***************************************************************************/
+
+    public void reinit ( )
+    {
+        this.root = eb_root.init;
+        this.root.unique = true;
+    }
+
+    /***************************************************************************
+
+        Adds a new node to the map for key or obtains the existing if already in
+        the map.
+
+        Params:
+            key   = the key to add a new node for or obtain the existing one
+            added = outputs true a new node was added or false if key was found
+                    in the map so that the existing node is returned
+
+        Returns:
+            the node associated with key. If added is true then it was newly
+            created and added, otherwise it is the already existing node in the
+            tree.
+
+    ***************************************************************************/
+
+    public Node* put ( ulong key, out bool added )
+    {
+        return this.put_(key, added);
+    }
+
+    /***********************************************************************
+
+        Looks up the node for key in the map.
+
+        Params:
+            key = the key to look up node for
+
+        Returns:
+            the node associated with key or null if not found.
+
+    ***********************************************************************/
+
+    public Node* opIn_r ( ulong key )
+    {
+        return cast(Node*)eb64_lookup(&this.root, key);
+    }
+
+    /***********************************************************************
+
+        Obtains the first or last node in the map.
+
+        Params:
+            first = `true`: obtain the first node,
+                    `false`: obtain the last node
+
+        Returns:
+            the first or last node or `null` if the map is empty.
+
+    ***********************************************************************/
+
+    public Node* getBoundary ( bool first = true ) ( )
+    {
+        static if (first)
+            alias eb64_first eb64_function;
+        else
+            alias eb64_last eb64_function;
+
+        return cast(Node*)eb64_function(&this.root);
+    }
+
+    /***********************************************************************
+
+        foreach iterator over nodes in the tree. Any tree modification is
+        permitted during iteration.
```

Comment:
This is actually a big advantage of this container, too. Worth mentioning in the relnotes / module header.